ui/src/main/java/edu/wpi/grip/ui/preview/ContoursSocketPreviewView.java

```diff
@@ -0,0 +1,84 @@
+package edu.wpi.grip.ui.preview;
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
+import edu.wpi.grip.core.OutputSocket;
+import edu.wpi.grip.core.events.SocketChangedEvent;
+import edu.wpi.grip.core.operations.composite.ContoursReport;
+import edu.wpi.grip.ui.util.ImageConverter;
+import javafx.application.Platform;
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.VBox;
+
+import static org.bytedeco.javacpp.opencv_core.*;
+import static org.bytedeco.javacpp.opencv_imgproc.drawContours;
+
+/**
+ * A preview view for displaying contours.  This view shows each contour as a different-colored outline (so they can be
+ * individually distinguished), as well as a count of the total number of contours found.
+ */
+public final class ContoursSocketPreviewView extends SocketPreviewView<ContoursReport> {
+
+    private final ImageConverter imageConverter = new ImageConverter();
+    private final ImageView imageView = new ImageView();
+    private final Label infoLabel = new Label();
+    private final Mat tmp = new Mat();
+
+    private final static Scalar[] CONTOUR_COLORS = new Scalar[]{
+            Scalar.RED,
+            Scalar.YELLOW,
+            Scalar.GREEN,
+            Scalar.CYAN,
+            Scalar.BLUE,
+            Scalar.MAGENTA,
+    };
+
+    /**
+     * @param eventBus The EventBus used by the application
+     * @param socket   An output socket to preview
+     */
+    public ContoursSocketPreviewView(EventBus eventBus, OutputSocket<ContoursReport> socket) {
+        super(eventBus, socket);
+
+        this.setContent(new VBox(this.imageView, this.infoLabel));
+        this.render();
+    }
+
+    @Subscribe
+    public void onSocketChanged(SocketChangedEvent event) {
+        if (event.getSocket() == this.getSocket()) {
+            this.render();
+        }
+    }
+
+    private void render() {
+        synchronized (this) {
+            final ContoursReport contours = this.getSocket().getValue().get();
+            long numContours = 0;
+
+            if (!contours.getContours().isNull() && contours.getRows() > 0 && contours.getCols() > 0) {
+                // Allocate a completely black OpenCV Mat to draw the contours onto.  We can easily render contours
+                // by using OpenCV's drawContours function and converting the Mat into a JavaFX Image.
+                this.tmp.create(contours.getRows(), contours.getCols(), CV_8UC3);
+                bitwise_xor(tmp, tmp, tmp);
+
+                numContours = contours.getContours().size();
+
+                for (int i = 0; i < numContours; i++) {
+                    drawContours(this.tmp, contours.getContours(), i, CONTOUR_COLORS[i % CONTOUR_COLORS.length]);
```

Comment:
No, I don't.

OpenCV returns contorus in order from top left to bottom right, and we color them according to the order that OpenCV returns them.  If you give it a really noise stream of images, the contours will change a lot, and it'll be hard to track them, and you'll be able to tell that.  If we color by size, you wouldn't be able to distinguish the individual contours, and it would be difficult for users to distinguish how well their pipeline is working and what needs to change in it.

As as simple example, this:

![a](https://cloud.githubusercontent.com/assets/3964980/11610508/a6349770-9b72-11e5-9cb8-96b6d35f9e46.png)

... is more informative than this...

![b](https://cloud.githubusercontent.com/assets/3964980/11610509/a8862c8c-9b72-11e5-9677-5f9bc9be7874.png)

Although in practice, coloring all contours with a size range the same will probably have this effect on noisy images:

![c](https://cloud.githubusercontent.com/assets/3964980/11610523/0e57b7b0-9b73-11e5-8fac-0dd18d63c503.png)